private/bufpkg/bufplugin/bufplugindocker/docker.go

```diff
@@ -0,0 +1,203 @@
+	Build(ctx context.Context, dockerfile io.Reader, config *bufpluginconfig.Config, options ...BuildOption) (*BuildResponse, error)
+	// Push the Docker image to the remote registry.
+	Push(ctx context.Context, image string, auth RegistryAuth) (*PushResponse, error)
+	// Delete removes the Docker image from local Docker Engine.
+	Delete(ctx context.Context, image string) (*DeleteResponse, error)
+	// Close releases any resources used by the underlying Docker client.
+	Close() error
+}
+
+type BuildOption func(*buildParams)
+
+func WithConfigDirPath(path string) BuildOption {
+	return func(params *buildParams) {
+		params.ConfigDirPath = path
+	}
+}
+
+type BuildResponse struct {
+	// Image contains the Docker image name in the local Docker engine including the tag (i.e. plugins.buf.build/library/some-plugin:<id>, where <id> is a random id).
+	// It is created from the bufpluginconfig.Config's Name.IdentityString() and a unique id.
+	Image string
+	// Digest specifies the Docker image digest in the format <hash_algorithm>:<hash>.
+	// Example: sha256:65001659f150f085e0b37b697a465a95cbfd885d9315b61960883b9ac588744e
+	Digest string
+}
+
+type PushResponse struct{}
+
+type DeleteResponse struct{}
+
+type dockerAPIClient struct {
+	cli    *client.Client
+	logger *zap.Logger
+}
+
+var _ Client = (*dockerAPIClient)(nil)
+
+func (d *dockerAPIClient) Build(ctx context.Context, dockerfile io.Reader, pluginConfig *bufpluginconfig.Config, options ...BuildOption) (*BuildResponse, error) {
+	params := &buildParams{}
+	for _, option := range options {
+		option(params)
+	}
+	// TODO: Need to determine how contextDir parameter is used in Docker engine.
+	buildkitSession, err := createSession(ctx, fmt.Sprintf("%s/%s", pluginConfig.Name.Owner(), pluginConfig.Name.Plugin()), params.ConfigDirPath, zap.L())
+	if err != nil {
+		return nil, err
+	}
+
+	dockerContext, err := createDockerContext(dockerfile)
+	if err != nil {
+		return nil, err
+	}
+
+	eg, errGroupCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		return buildkitSession.Run(errGroupCtx, func(ctx context.Context, proto string, meta map[string][]string) (net.Conn, error) {
+			return d.cli.DialHijack(ctx, "/session", proto, meta)
+		})
+	})
+
+	buildID := stringid.GenerateRandomID()
+	imageName := pluginConfig.Name.IdentityString() + ":" + buildID
+	eg.Go(func() error {
+		defer buildkitSession.Close()
+
+		response, err := d.cli.ImageBuild(ctx, dockerContext, types.ImageBuildOptions{
+			Tags:     []string{imageName},
+			Platform: "linux/amd64",
+			Labels: map[string]string{
+				"build.buf.plugins.config.remote": pluginConfig.Name.Remote(),
+				"build.buf.plugins.config.owner":  pluginConfig.Name.Owner(),
+				"build.buf.plugins.config.name":   pluginConfig.Name.Plugin(),
+			},
+			Version:   types.BuilderBuildKit, // DOCKER_BUILDKIT=1
+			SessionID: buildkitSession.ID(),
+		})
+		if err != nil {
+			return err
+		}
+		defer response.Body.Close()
```

Comment:
We should capture these `Close` errors in a deferred function (the same applies to the other deferred function in this `eg.Go` call above):
```suggestion
		defer func() {
		    retErr = multierr.Append(retErr, response.Body.Close())
		}()
```